Add functionality to pull sample type

chantelwetzel-noaa opened the pull request:
These changes do the following:

1. Add a function argument to pull_catch and PullCatch.fn that allows users to pull all partition sample type data. The argument documentation notes that this primarily pertains to Pacific hake.
2. Create a new function that allows users to combine catch data from the same tow which can occur if there are multiple partition sample types on a single tow.  The pull_catch function adds a warning comment but does not augment the data.
3.  Add tests for pulling data with all partition sample types.

chantelwetzel-noaa commented:
@kellijohnson-NOAA I wanted to check in to make sure you received an email for review.  This pull request handles pulling all sample-type data that is particularly specific to Pacific hake. Let me know if you have questions or do not have time for review.

kellijohnson-NOAA commented on R/pull_catch.R:
```diff
@@ -262,17 +259,25 @@ pull_catch <- function(common_name = NULL,
+  if(sum(c("Life Stage", "Size") %in% sample_types) == 2) {
+    n_id <- table(catch$trawl_id)
+    if(sum(n_id == 2) > 0){
+      print("Warning: Pulling all sample types (Life Stage and Size) has resulted in multiple records for unique tows (Trawl_id).")
+      print("The `combine_tows` function can be used to combine these multiple records for unique tows if needed.")
```
The function call in the test leads to
![image](https://github.com/pfmc-assessments/nwfscSurvey/assets/4108564/4ae28bf9-0060-4d40-a0c6-0db50f354b65)
because you are using `print()` rather than `warning()`. I am not sure if that was intentional or not but print messages can be missed when using markdown and other ways of running R, where `message()`, `warning()`, `error()`, or calls from the `cli` package can sometimes be a more robust way to communicate information.

chantelwetzel-noaa replied:
Thank you for catching this. I used print because that is my mental default but revised to warning.

kellijohnson-NOAA commented on R/combine_tows.R:
```diff
@@ -0,0 +1,58 @@
+#'
+#'
+#'
+combine_tows <- function(dir = NULL, data, verbose = TRUE){
```
I think that `data` should be the first argument rather than `dir` because there is no default for data.

kellijohnson-NOAA commented on R/combine_tows.R:
```diff
@@ -0,0 +1,58 @@
+#' Combine catch data by trawl_id for data pulled using the
```
For all "code" in the roxygen, you can use backticks to signify that it is code, e.g., `trawl_id`.

kellijohnson-NOAA replied:
The first line should be 60 characters giving a brief title of the function without a full stop followed by a blank line. Then the description of the function should start on line three and can be a paragraph.

kellijohnson-NOAA commented on R/combine_tows.R:
```diff
@@ -0,0 +1,58 @@
+#' Combine catch data by trawl_id for data pulled using the
+#' pull_catch function. Generally, only a single tow is returned
+#' by species and in those instances this function is not needed.
+#' This function is only needed in unique data situations. For
```
The sentence starting with "This" is repetitive with the previous sentence.

kellijohnson-NOAA commented on R/combine_tows.R:
```diff
@@ -0,0 +1,58 @@
+#' pull_catch function. Generally, only a single tow is returned
+#' by species and in those instances this function is not needed.
+#' This function is only needed in unique data situations. For
+#' example, if data are pulled with all sample_types included then
```
There should be a comma before "then".

kellijohnson-NOAA commented on R/combine_tows.R:
```diff
@@ -0,0 +1,58 @@
+#' This function is only needed in unique data situations. For
+#' example, if data are pulled with all sample_types included then
+#' there will often be data from the same trawl_id included as
+#' multiple records. In these instances this function allows for
```
There should be a comma after "instances".

kellijohnson-NOAA commented on R/combine_tows.R:
```diff
@@ -0,0 +1,58 @@
+#' example, if data are pulled with all sample_types included then
+#' there will often be data from the same trawl_id included as
+#' multiple records. In these instances this function allows for
+#' the sample data to be combined for unique trawl_id.
```
Add "each" before "unique".

kellijohnson-NOAA commented on R/combine_tows.R:
```diff
@@ -0,0 +1,58 @@
+
+  check_dir(dir = dir, verbose = verbose)
+
+  if(!"total_catch_numbers" %in% colnames(data)){
```
All if statements should be followed by a space and there should be a space before the curly bracket. Same on line 36.

chantelwetzel-noaa replied:
Thank you for catching this code formatting

kellijohnson-NOAA commented on R/combine_tows.R:
```diff
@@ -0,0 +1,58 @@
+
+  catch <- data |>
+    dplyr::group_by(common_name, trawl_id) |>
+    mutate(
```
You didn't import `mutate` so it should be `dplyr::mutate()` but I am uncertain why you are using mutate? I would have thought that you would have used `dplyr::summarize()` and then you would not need to use `dplyr::distinct()`.

chantelwetzel-noaa replied:
I used mutate intentionally since I want a data frame with all the original columns returned but only the values modified in a select few adjusted.

kellijohnson-NOAA commented on R/combine_tows.R:
```diff
@@ -0,0 +1,58 @@
+      subsample_wt_kg = sum(subsample_wt_kg),
+      cpue_kg_per_ha_der = sum(cpue_kg_per_ha_der),
+      cpue_kg_km2 = sum(cpue_kg_km2),
+      partition_sample_types = NA,
```
Wouldn't you want to combine the entries into a single new entry using `paste()` rather than assign them as `NA`? Same comment for the line below as well.

kellijohnson-NOAA replied:
Thanks @chantelwetzel-noaa this is really informative now because users can see where Size samples were combined with Size samples and so on.

kellijohnson-NOAA commented on man-roxygen/sample_types.R:
```diff
@@ -0,0 +1,10 @@
+#' Pacific hake and should not be considered different than regular survey
+#' samples. The other types of samples that may be designated Life Stage are
+#' egg cases that can be caught and identified for select elasmobranch
+#' species. These type of samples would not be included in the data to estimate
```
Change "would" to "should". And, add "used" after "data".

chantelwetzel-noaa replied:
Revised

kellijohnson-NOAA commented on tests/testthat/test-data.R:
```diff
@@ -57,6 +57,30 @@ test_that("pull_catch-multispecies", {
+  combine_hake <- combine_tows(
+    data = data_hake
+  )
+  expect_equal(length(unique(data_hake$Trawl_id)), nrow(combine_hake))
```
I am not sure how rigorous you would like the tests to be but it seems like there should be some test only using the first three sample types and seeing if you get the same number of rows as when you pull all four sample types but omit the fourth type after pulling. And, that some of the weights are added correctly. Also, that if you pull all sample types for a species without Life Stage samples that you get the same number of samples as when you use the default value for sample_type. Another test would be to pull a species with weird egg samples and make sure that they do not make it into the final data that is returned.

chantelwetzel-noaa replied:
I have added additional tests that check data pulled with partition = eggs and compare pulls with all sample types versus a subset.

kellijohnson-NOAA reviewed (changes requested):
Sorry for the delay @chantelwetzel-noaa. I made some rough comments and would like to review it again after you respond. Two major things, I think that there should be more test written and that there should not be changes made to the deprecated function.

chantelwetzel-noaa commented on tests/testthat/test-data.R:
```diff
@@ -57,6 +57,30 @@ test_that("pull_catch-multispecies", {
+test_that("pull-sample-types", {
+  skip_on_cran()
+
+  set.seed(123)
```
Whoever created the original tests for this included the set.seed and I have been copying ever since but not understanding why it was needed.  I have removed the set.seed function throughout this script

chantelwetzel-noaa commented on R/PullCatch.fn.R:
You are correct.  This addition was not needed and has been removed.

chantelwetzel-noaa commented:
@kellijohnson-NOAA I think I have addressed all the items that you identified in your initial review.  I also added code to the `combine_tows` function to pull out samples that should not be used for index creation and code to save the modified data frame if dir != NULL.

kellijohnson-NOAA commented:
Great, thank you so much @chantelwetzel-noaa for being so speedy. I will start this tonight but I probably won't have time to finish. But, tomorrow I have time.

kellijohnson-NOAA commented on R/combine_tows.R:
```diff
@@ -0,0 +1,73 @@
+
+  find <- grep("trawl_id", colnames(data), ignore.case = TRUE)
+  n_id <- table(data[, find])
+  if (sum(n_id == 2) == 0) {
```
Would there ever be an instance where there are three or more? I think that a more robust test would be
```
if (all(n_id == 1)) {
```

kellijohnson-NOAA commented on R/pull_catch.R:
```diff
@@ -262,17 +259,25 @@ pull_catch <- function(common_name = NULL,
 
+  if(sum(c("Life Stage", "Size") %in% sample_types) == 2) {
+    n_id <- table(catch$trawl_id)
+    if(sum(n_id == 2) > 0){
```
Same comment as above regarding a more robust test
```
if (any(n_id > 1)) {

chantelwetzel-noaa replied:
That is a great suggestion.  Thank you!

kellijohnson-NOAA reviewed (approved):
I made two additional comments but you can merge the PR in without addressing them if you want.
